ministryofjustice/laa-apply-for-criminal-legal-aid
# Submission end to end with datastore (feature flag)

## Description of change
This is a first stab at an integration between Apply and Datastore for submitting applications. An end to end so to speak from a provider's point of view. Note however there is no proper adapter to create the JSON document yet, and I'm using a dirty, quick sampling of attributes.

It currently is behind a feature flag, disabled on local by default as to not disrupt any work or force developers to run a local datastore service (for now).

Also disabled on staging until there is something more solid.

This uses a little [http client gem](https://github.com/ministryofjustice/laa-criminal-applications-datastore-api-client) made for the only purpose of wrapping the http API in a common interface that could be shared between Apply and Review eventually.

## Link to relevant ticket
Follow-up to datastore spike ticket to continue validation of assumptions and data model:
https://dsdmoj.atlassian.net/browse/CRIMAP-202

## How to manually test the feature
You can test the feature as follows:

If you just check out the branch, the feature is turned off by default. Edit the `settings.yml` and turn on the feature. If you do so, you will need to have the datastore service ([repo here](https://github.com/ministryofjustice/laa-criminal-applications-datastore)) running locally. You can use docker-compose for this.

By default the local datastore would be exposed on `http://localhost:3003` but if that is not the case, or if you made changes to suit your preferences, then in your `.env.development.local` expose the endpoint as `DATASTORE_API_ROOT=http://localhost:3003`

Restart the apply server to pick up any changes to feature flags or env variables. Fill in an application all the way to the submission and observe it will be, in addition to the local database, be pushed to the datastore.
You can check it is there by then going to a browser (or using Postman) and retrieving all applications with:
`http://localhost:3003/api/v1/applications`
